src/flb_upstream.c

```diff
@@ -791,10 +756,44 @@ struct flb_connection *flb_upstream_conn_get(struct flb_upstream *u)
     }
 
     /*
-     * There are no keepalive connections available or keepalive is disabled
-     * so we need to create a new one.
+     * If no reusable connection is found, then check the total connection limit
+     * and create a new connection if allowed.
      */
+
     if (conn == NULL) {
+    /* If the upstream is limited by max connections, check current state */
+        if (u->base.net.max_worker_connections>0) {
+            /*
+             * Connections are linked to one of these lists:
+             *
+             *  - av_queue  : connections ready to be used (available)
+             *  - busy_queue: connections that are busy (someone is using them)
+             *  - drop_queue: connections in the cleanup phase (to be drop)
+             *
+             * Fluent Bit doesn't create connections ahead of time, just on demand. When
+             * a connection is created is placed into the busy_queue, when it is no longer
+             * needed one of these things happen:
+             *
+             *   - if keepalive is enabled (default), the connection is moved to the 'av_queue'.
+             *   - if keepalive is disabled, the connection is moved to 'drop_queue' then is
+             *     closed and destroyed.
+             *
```

Comment:
note: if a connection presented an error is always moved to `destroy_queue` (with or without keepalive enabled)